experimental/oak_baremetal_channel/src/lib.rs

```diff
@@ -66,16 +105,233 @@ impl Frame {
+        );
+
+        let length_bytes = frame_length.to_le_bytes();
+        frame_bytes.extend_from_slice(&length_bytes);
+        let method_or_status_bytes = frame.method_or_status.to_le_bytes();
+        frame_bytes.extend_from_slice(&method_or_status_bytes);
+        let padding_bytes = [0; PADDING_SIZE];
+        frame_bytes.extend_from_slice(&padding_bytes);
+        let flag_bytes: [u8; FLAG_SIZE] = frame.flag.into();
+        frame_bytes.extend_from_slice(&flag_bytes);
+        frame_bytes.extend_from_slice(&frame.body);
+
+        Ok(frame_bytes)
+    }
+}
+
+/// A [`Message`] is conceptual unit of data that is sent over the communication
+/// channel. It can represent either a full request, or full response of an oak_idl
+/// invocation. Requests and responses can be converted to a [`Message`], and
+/// vice versa. Usually a single [`Message`] is sent over the communication
+/// channel over several [`Frame`]s.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Message {
+    pub method_or_status: MethodOrStatus,
+    pub body: Vec<u8>,
+}
+
+/// Deconstruct a [`Message`] into a set of [`Frame`]s.
+impl From<Message> for Vec<Frame> {
+    fn from(message: Message) -> Self {
+        let frame_bodies: Vec<Vec<u8>> = if message.body.is_empty() {
+            // Messages with an empty body are handled differently, since
+            // chunking an empty [`Vec`] would result in 0 chunks.
+            vec![vec![]]
+        } else {
+            message
+                .body
+                .chunks(MAX_FRAME_BODY_SIZE)
+                // TODO(#2848): It'd be nice if we didn't have to reallocate here.
+                // We may be able to avoid doing so by making [`Frame`] use
+                // slices and lifetimes. Or alternatively use the Bytes crate for
+                // reference counting.
+                .map(|frame_body| frame_body.to_vec())
+                .collect()
+        };
+
+        let number_of_frames = frame_bodies.len();
+        frame_bodies
+            .into_iter()
+            .enumerate()
+            .map(|(index, body)| {
+                let flag = {
+                    if index == 0 {
+                        match number_of_frames {
+                            1 => Flag::AtomicMessage,
+                            _ => Flag::MessageStart,
+                        }
+                    } else if index == number_of_frames - 1 {
+                        Flag::MessageMessage
+                    } else {
+                        Flag::MessageContinuation
+                    }
+                };
+                Frame {
+                    method_or_status: message.method_or_status,
+                    flag,
+                    body,
+                }
+            })
+            .collect()
+    }
+}
+
+/// Struct to incrementally build a [`Message`] from a set of [`Frame`]s.
+#[derive(Default, Debug, PartialEq, Eq)]
+struct PartialMessage {
+    inner: Option<Message>,
+}
```

Comment:
I'm not sure I see the value of building a partial message and having an inner message which is semantically not a message. How about just keeping track of the Vec<Vec<u8>> parts here (i.e. just append the next frame each time), and once all of them are collected, then just do a single new allocation of a new Vec to concat all of them into a single final body (which is the message body). This way there is never a malformed intermediate Message (which IMO currently makes things semantically unclear), and also there is a single allocation instead of potentially many intermediate reallocations.
Also the add_frame method may be simplified (though I'm not sure to what extent).